File: ReactAndroid/src/main/java/com/facebook/react/animated/ColorAnimatedNode.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.animated;

import android.content.Context;
import android.graphics.Color;
import android.view.View;
import com.facebook.react.bridge.ColorPropConverter;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.views.view.ColorUtil;

/** Animated node that represents a color. */
/*package*/ class ColorAnimatedNode extends AnimatedNode {

  private final NativeAnimatedNodesManager mNativeAnimatedNodesManager;
  private final ReactApplicationContext mReactApplicationContext;
  private final int mRNodeId;
  private final int mGNodeId;
  private final int mBNodeId;
  private final int mANodeId;
  private final ReadableMap mNativeColor;
  private boolean mNativeColorApplied;

  public ColorAnimatedNode(
      ReadableMap config,
      NativeAnimatedNodesManager nativeAnimatedNodesManager,
      ReactApplicationContext reactApplicationContext) {
    mNativeAnimatedNodesManager = nativeAnimatedNodesManager;
    mReactApplicationContext = reactApplicationContext;
    mRNodeId = config.getInt("r");
    mGNodeId = config.getInt("g");
    mBNodeId = config.getInt("b");
    mANodeId = config.getInt("a");
    mNativeColor = config.getMap("nativeColor");
    tryApplyNativeColor();
  }

  public int getColor() {
    tryApplyNativeColor();

    ValueAnimatedNode rNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mRNodeId);
    ValueAnimatedNode gNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mGNodeId);
    ValueAnimatedNode bNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mBNodeId);
    ValueAnimatedNode aNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mANodeId);

    double r = rNode.getValue();
    double g = gNode.getValue();
    double b = bNode.getValue();
    double a = aNode.getValue();

    return ColorUtil.normalize(r, g, b, a);
  }

  @Override
  public String prettyPrint() {
    return "ColorAnimatedNode["
        + mTag
        + "]: r: "
        + mRNodeId
        + " g: "
        + mGNodeId
        + " b: "
        + mBNodeId
        + " a: "
        + mANodeId;
  }

  private void tryApplyNativeColor() {
    if (mNativeColor == null || mNativeColorApplied) {
      return;
    }

    Context context = getContext();
    if (context == null) {
      return;
    }

    int color = ColorPropConverter.getColor(mNativeColor, context);

    ValueAnimatedNode rNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mRNodeId);
    ValueAnimatedNode gNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mGNodeId);
    ValueAnimatedNode bNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mBNodeId);
    ValueAnimatedNode aNode = (ValueAnimatedNode) mNativeAnimatedNodesManager.getNodeById(mANodeId);

    rNode.mValue = Color.red(color);
    gNode.mValue = Color.green(color);
    bNode.mValue = Color.blue(color);
    aNode.mValue = Color.alpha(color) / 255.0;

    mNativeColorApplied = true;
  }

  private Context getContext() {
    Context context = mReactApplicationContext.getCurrentActivity();
    if (context != null) {
      return context;
    }

    // There are cases where the activity may not exist (such as for VRShell panel apps). In this
    // case we will search for a view associated with a PropsAnimatedNode to get the context.
    return getContextHelper(this);
  }

  private static Context getContextHelper(AnimatedNode node) {
    // Search children depth-first until we get to a PropsAnimatedNode, from which we can
    // get the view and its context
    if (node.mChildren != null) {
      for (AnimatedNode child : node.mChildren) {
        if (child instanceof PropsAnimatedNode) {
          View view = ((PropsAnimatedNode) child).getConnectedView();
          return view != null ? view.getContext() : null;
        } else {
          return getContextHelper(child);
        }
      }
    }
    return null;
  }
}

File: packages/react-native/ReactCommon/react/renderer/animated/nodes/ColorAnimatedNode.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/*
 * Adapted from react-native-windows under the MIT license.
 */

#include "ColorAnimatedNode.h"
#include <react/renderer/animated/NativeAnimatedNodesManager.h>
#include <react/renderer/animated/nodes/ValueAnimatedNode.h>

namespace facebook::react {

namespace {

uint8_t getColorValue(
    const NativeAnimatedNodesManager& manager,
    Tag nodeTag,
    bool isDecimal = false) {
  if (const auto node = manager.getAnimatedNode<ValueAnimatedNode>(nodeTag)) {
    if (isDecimal) {
      return std::clamp(
          static_cast<uint32_t>(node->getValue() * 255), 0u, 255u);
    } else {
      return std::clamp(static_cast<uint32_t>(node->getValue()), 0u, 255u);
    }
  }
  return 0;
}

uint8_t getAlphaValue(const NativeAnimatedNodesManager& manager, Tag nodeTag) {
  return getColorValue(manager, nodeTag, true);
}

} // namespace

ColorAnimatedNode::ColorAnimatedNode(
    Tag tag,
    const folly::dynamic& config,
    NativeAnimatedNodesManager& manager)
    : AnimatedNode(tag, config, manager, AnimatedNodeType::Color),
      rNodeTag_(static_cast<Tag>(getConfig()["r"].asInt())),
      gNodeTag_(static_cast<Tag>(getConfig()["g"].asInt())),
      bNodeTag_(static_cast<Tag>(getConfig()["b"].asInt())),
      aNodeTag_(static_cast<Tag>(getConfig()["a"].asInt())) {}

void ColorAnimatedNode::update() {
  color_ = *colorFromRGBA(
      getColorValue(*manager_, rNodeTag_),
      getColorValue(*manager_, gNodeTag_),
      getColorValue(*manager_, bNodeTag_),
      getAlphaValue(*manager_, aNodeTag_));
}

Color ColorAnimatedNode::getColor() {
  if (manager_->updatedNodeTags_.contains(tag_)) {
    update();
    manager_->updatedNodeTags_.erase(tag_);
  }
  return color_;
  return 0;
}

} // namespace facebook::react
